codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithygo/codegen/StructureGenerator.java

```diff
@@ -117,134 +113,118 @@ public void renderStructure(Runnable runnable, boolean isInputStructure) {
-    // as document types themselves, or part of broader document-type structures. This avoids making backwards
-    // incompatible changes if the document type representation changes if it is later annotated as a modeled
-    // document type. This restriction may be relaxed later by removing this constraint.
-    writer.write("");
-    writer.write("$L", ProtocolDocumentGenerator.NO_DOCUMENT_SERDE_TYPE_NAME);
-
     writer.closeBlock("}").write("");
+    validationGenerator.renderValidator(shape, isInputStructure);
   }
 
   /**
    * Renders an error structure and supporting methods.
    */
   private void renderErrorStructure() {
     Symbol structureSymbol = symbolProvider.toSymbol(shape);
-    writer.addUseImports(SmithyGoDependency.SMITHY);
     writer.addUseImports(SmithyGoDependency.FMT);
     ErrorTrait errorTrait = shape.expectTrait(ErrorTrait.class);
 
     // Write out a struct to hold the error data.
-    writer.writeShapeDocs(shape);
     writer
       .openBlock(
         "type $L struct {",
         "}",
         structureSymbol.getName(),
         () -> {
-          // The message is the only part of the standard APIError interface that isn't known ahead of time.
-          // Message is a pointer mostly for the sake of consistency.
-          writer.write("Message *string").write("");
-          writer.write("ErrorCodeOverride *string").write("");
-
+          writer.write(
+            "$LBaseException",
+            context.settings().getService().getName()
+          );
+          Set<String> memberNameSet = new HashSet<>();
+          // TODO: Revisit if message has to be strictly pointer or not (even with required trait).
+          // When any shape is required we don't add pointer in local service but AWS SDK does.
           for (MemberShape member : shape.getAllMembers().values()) {
             String memberName = symbolProvider.toMemberName(member);
-            // error messages are represented under Message for consistency
-            if (!ERROR_MEMBER_NAMES.contains(memberName)) {
-              writer.write(
-                "$L $P",
-                memberName,
-                symbolProvider.toSymbol(member)
-              );
-            }
+            memberNameSet.add(memberName);
+            writer.write("$L $P", memberName, symbolProvider.toSymbol(member));
           }
 
-          writer.write("");
-          writer.write(
-            "$L",
-            ProtocolDocumentGenerator.NO_DOCUMENT_SERDE_TYPE_NAME
-          );
+          // The message is the only part of the standard APIError interface that isn't known ahead of time.
+          // Message is a pointer mostly for the sake of consistency.
+
+          // If Message and ErrorCodeOverride is not defined in model.
+          if (!memberNameSet.contains("Message")) {
+            writer.write("Message *string").write("");
+          }
+          if (!memberNameSet.contains("ErrorCodeOverride")) {
```

Comment:
(Curious) Is this like "exception type"?

Reply:
This is to keep it consistent with the smithy-go convention.